### PR TITLE
Cut genericapserver->registry dependency

### PR DESCRIPTION
**Pull request by sttts**
Towards moving genericapiserver into staging.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37547)

### CODE REVIEW

File: pkg/registry/generic/options.go [hunk 1]
```diff
@@ -16,7 +16,11 @@ limitations under the License.
 
 package generic
 
-import "k8s.io/kubernetes/pkg/storage/storagebackend"
+import (
+	"k8s.io/kubernetes/pkg/genericapiserver"
```
Comment (sttts) [inline pkg/registry/generic/options.go]: I mean the interfaces.

Comment (sttts) [inline pkg/registry/generic/options.go]: Then let's just move them to master.

Comment (sttts) [inline pkg/registry/generic/options.go]: A dependency from pkg/registry back to master is wrong as well. I have moved the two interfaces now to pkg/registry. I think that's the right place. @deads2k ptal

**Comment by sttts** [comment]
@k8s-bot cvm gce e2e test this

**Comment by sttts** [comment]
@deads2k ptal ^^

File: pkg/master/master.go [hunk 2]
```diff
@@ -43,14 +43,14 @@ import (
...
+	"k8s.io/kubernetes/pkg/routes"
 	"k8s.io/kubernetes/pkg/runtime/schema"
 	utilnet "k8s.io/kubernetes/pkg/util/net"
 	nodeutil "k8s.io/kubernetes/pkg/util/node"
 
-	"k8s.io/kubernetes/pkg/registry/generic"
-	"k8s.io/kubernetes/pkg/registry/generic/registry"
-	"k8s.io/kubernetes/pkg/routes"
-
 	"github.com/golang/glog"
```
Comment (deads2k) [inline pkg/master/master.go]: while you're reorganizing imports, I'm inclined to follow openshift with 
 1. standard lib
 2. other deps
 3. kube deps

File: pkg/registry/interfaces.go [hunk 3]
```diff
@@ -14,16 +14,19 @@ See the License for the specific language governing permissions and
...
+	"k8s.io/kubernetes/pkg/genericapiserver"
 	"k8s.io/kubernetes/pkg/registry/generic"
 	"k8s.io/kubernetes/pkg/runtime/schema"
 )
 
+// RESTOptionsGetter is a function type for returning RESTOptions for a GroupVersion.
 type RESTOptionsGetter func(resource schema.GroupResource) generic.RESTOptions
 
+// RESTStorageProvider is a factory type for REST storage.
 type RESTStorageProvider interface {
```
Comment (deads2k) [inline pkg/registry/interfaces.go]: Only used in master right?  The others are just type assertions?

Comment (deads2k) [inline pkg/registry/interfaces.go]: > Only used in master right? The others are just type assertions?

If so, move to master and the map/list will keep them forced correct.

File: pkg/master/master.go (hunk 2 above)
Comment (sttts) [inline pkg/master/master.go]: Am fine with that order. But I would avoid any unnecessary changes here to escape rebase hell (as far as possible).

File: pkg/registry/interfaces.go (hunk 3 above)
Comment (sttts) [inline pkg/registry/interfaces.go]: don't like to link registry back to master. One direction is enough.

File: pkg/registry/interfaces.go [hunk 4]
```diff
@@ -14,16 +14,19 @@ See the License for the specific language governing permissions and
...
+package registry
 
 import (
+	"k8s.io/kubernetes/pkg/genericapiserver"
 	"k8s.io/kubernetes/pkg/registry/generic"
 	"k8s.io/kubernetes/pkg/runtime/schema"
 )
 
+// RESTOptionsGetter is a function type for returning RESTOptions for a GroupVersion.
 type RESTOptionsGetter func(resource schema.GroupResource) generic.RESTOptions
```
Comment (sttts) [inline pkg/registry/interfaces.go]: generic is more like a generic base impl, not generic types.

File: pkg/registry/interfaces.go (hunk 3 above)
Comment (deads2k) [inline pkg/registry/interfaces.go]: All the other uses are just type assertions, right?  You can delete them and our usage will keep them in line.

File: pkg/registry/interfaces.go (hunk 4 above)
Comment (sttts) [inline pkg/registry/interfaces.go]: I would prefer to keep those separate.

File: pkg/registry/interfaces.go (hunk 3 above)
Comment (sttts) [inline pkg/registry/interfaces.go]: no, we use it in `Master.InstallAPIs`

Comment (sttts) [inline pkg/registry/interfaces.go]: moved to master.

**Comment by deads2k** [comment]
lgtm, could use some squashing.

**Comment by deads2k** [comment]
squash bazel.  Its annoying, but there's no reason to splash the history

**Comment by sttts** [comment]
@k8s-bot gce etcd3 e2e test this
